RFC: Give users control over feature unification (#3692)
rust-lang/rfcs: text/3692-feature-unification.md

- [#8157: --bin B resolves features differently than -p B in a workspace](https://github.com/rust-lang/cargo/issues/8157)
- [#13844: The cargo build --bins re-builds binaries again after cargo build --all-targets](https://github.com/rust-lang/cargo/issues/13844)

# Motivation
[motivation]: #motivation

Today, when Cargo is building, features in dependencies are enabled baed on the set of packages selected to build.
This is an attempt to balance
- Build speed: we should reuse builds between packages within the same invocation
- Ability to verify features for a given package

This isn't always ideal.

If a user is building an application, they may be jumping around the application's components which are packages within the workspace.
The final artifact is the same but Cargo will select different features depending on which package they are currently building,
causing build churn for the same set of dependencies that, in the end, will only be used with the same set of features.
The "cargo-workspace-hack" is a pattern that has existed for years
(e.g. [`rustc-workspace-hack`](https://crates.io/crates/rustc-workspace-hack))

> where users have all workspace members that depend on a generated package that depends on direct-dependemncies in the workspace along with their features.

Reviewer: ```suggestion
where users have all workspace members that depend on a generated package that depends on direct-dependencies in the workspace along with their features.
```

- Introducing new named concepts.
- Explaining the feature largely in terms of examples.
- Explaining how Rust programmers should *think* about the feature, and how it should impact the way they use Rust. It should explain the impact as concretely as possible.
- If applicable, provide sample error messages, deprecation warnings, or migration guidance.
- If applicable, describe the differences between teaching this to existing Rust programmers and new Rust programmers.
- Discuss how this impacts the ability to read, understand, and maintain Rust code. Code is read and modified far more often than written; will the proposed feature make code easier to maintain?

For implementation-oriented RFCs (e.g. for compiler internals), this section should focus on how compiler contributors should think about the change, and give examples of its concrete impact. For policy RFCs, this section should provide an example-driven introduction to the policy, and explain its impact in concrete terms.

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

### Rust Version

> We'll add two new modes to feature unifcation:

Reviewer 2: ```suggestion
We'll add two new modes to feature unification:
```

**Unify features across the workspace, independent of the selected packages**

This would be built-in support for "cargo-workspace-hack".

This would require effectively changing from
1. Resolve dependencies
2. Filter dependencies down for current target and selected packages
3. Resolve features

To
1. Resolve dependencies
2. Filter dependencies down for current target
3. Resolve features
4. Filter for selected packages

> **Features will be evaluated for each package in isolation**

Author: Thanks for calling this out! I had considered this but forgot it and didn't want to hold it up until I could remember.

> **Features will be evaluated for each package in isolation**

Author: @sunshowers says `cargo-hakari` will unify normal and dev-dependencies.  Platforms are not unified.

> **Features will be evaluated for each package in isolation**

Author: @sunshowers says that is also unifies the features that workspace member features enable

> **Features will be evaluated for each package in isolation**

Author: Some cases with unifying dev-dependencies include
-`fail` being used for dev-dependencies but not production
- Tests enabling `std` but not wanting them for normal build
- Some use case related to private keys where the `impl Clone` is only provided on debug builds and not release builds

There are likely other cases of this same sort of problem

> **Features will be evaluated for each package in isolation**

Reviewer 3: In Miri's case it's dev-dependencies enabling different features for `libc` or some other fundamental crate, which means that even after `cargo test`, doing `cargo build` does an almost complete re-build of Miri -- that's waiting time I'd rather avoid. :)

> **Features will be evaluated for each package in isolation**

Author: I've summarized this under "Future possibilities"

> This will require building duplicate copies of build units when there is disjoint sets of features.

Reviewer 2: ```suggestion
This will require building duplicate copies of build units when there are disjoint sets of features.
```

- Feature Name: `feature-unification`
- Start Date: 2024-09-11
- RFC PR: [rust-lang/rfcs#3692](https://github.com/rust-lang/rfcs/pull/3692)
- Rust Issue: [rust-lang/rust#3692](https://github.com/rust-lang/rust/issues/3692)

# Summary
[summary]: #summary

> Allow users to control feature unifcation.

Reviewer 4: ```suggestion
Allow users to control feature unification.
```

and then locally patch in the real instance of it.

This also makes testing of features more difficult because a user can't just run `cargo check --workspace` to verify that the correct set of features are enabled.
This has led to the rise of tools like [cargo-hack](https://crates.io/crates/cargo-hack) which de-unify packages.

# Guide-level explanation
[guide-level-explanation]: #guide-level-explanation

# Reference-level explanation
[reference-level-explanation]: #reference-level-explanation

We'll add two new modes to feature unification:

**Unify features across the workspace, independent of the selected packages**

This would be built-in support for "cargo-workspace-hack".

This would require effectively changing from
1. Resolve dependencies
2. Filter dependencies down for current target and selected packages
3. Resolve features

To
1. Resolve dependencies
2. Filter dependencies down for current target
3. Resolve features
4. Filter for selected packages

**Features will be evaluated for each package in isolation**

This will require building duplicate copies of build units when there are disjoint sets of features.

> For example purposes., this could be implemented as either

Reviewer 4: ```suggestion
For example, this could be implemented as either
```

- Loop over the packages, resolving, and then run a build plan for that package
- Resolve for each package and generate everything into the same build plan

This is not prescriptive of the implementation but to illustrate what the feature does.
The initial implementation may be sub-optimal.
Likely, the implementation could be improved over time.

**Note:** these features do not need to be stabilized together.

##### `resolver.feature-unification`

*(update to [Configuration](https://doc.rust-lang.org/cargo/reference/config.html))*

* Type: string
* Default: "selected"
* Environment: `CARGO_RESOLVER_FEATURE_UNIFICATION`

Specify which packages participate in [feature unification](https://doc.rust-lang.org/cargo/reference/features.html#feature-unification).

* `selected`: merge dependency features from all package specified for the current build
* `workspace`: merge dependency features across all workspace members, regardless of which packages are specified for the current build

> * `package`: dependency features are only enabled for each package, preferring duplicate builds of dependencies to when different feature sets are selected

Reviewer 4: ```suggestion
* `package`: dependency features are only enabled for each package, preferring duplicate builds of dependencies when different feature sets are selected by their dependents
```

# Drawbacks
[drawbacks]: #drawbacks

This increases entropy within Cargo and the universe at large.

If someone enables mutually exclusive features in different packages, then `workspace` unification will fail.
Officially, features are supposed to be additive, making mutually exclusive features officially unsupported.
Instead, effort should be put towards [official mutually exclusive globals](https://internals.rust-lang.org/t/pre-rfc-mutually-excusive-global-features/19618).

Some features cannot be enabled in some packages, like a `no_std` package not wanting `std` features.
These workspaces will not be able to use `workspace` unification.
For now, unifying for the `"workspace"` is primarily targeted at single-application workspaces.
The other config fields can always be used instead.

# Rationale and alternatives
[rationale-and-alternatives]: #rationale-and-alternatives

This is done in the config instead of the manifest:
- As this can change from run-to-run, this covers more use cases

> - As this fits easily into the `resolver` table. there is less design work

Author: @dlight 

> Could this be added into the workspace's `Cargo.toml` rather than `.cargo/config.toml`? Or is this intended to be applied to all workspaces inside a directory?

*(moved here for easier to follow thread)*

Author: We do mention manifest support in [future possibilities](https://github.com/epage/rfcs/blob/feature-unification/text/3692-feature-unification.md#manifest-support).

Also a config doesn't  require it to apply to all workspaces.  It depends on where you put your config, where your workspaces are, and what directory you are in.

> - As this can change from run-to-run, this covers more use cases
> - As this fits easily into the `resolver` table. there is less design work

Reviewer 4: ```suggestion
- As this can change from run to run, this covers more use cases.
- As this fits easily into the `resolver` table, there is less design work.
```

> We could extend this with configuration for to exclude packages for the various use cases mentioned.
> Supporting excludes as environemnt/project configuration complexity as well as implementation complexity.

Reviewer 4: ```suggestion
We could extend this with configuration to exclude packages for the various use cases mentioned.
Supporting excludes adds environment/project configuration complexity as well as implementation complexity.
```

# Prior art
[prior-art]: #prior-art

[`cargo-hakari`](https://crates.io/crates/cargo-hakari) is a "cargo-workspace-hack" generator that builds a graph off of `cargo metadata` and re-implements feature unification.

[cargo-hack](https://crates.io/crates/cargo-hack) can run each selected package in a separate `cargo` invocation to prevent unification.

# Unresolved questions
[unresolved-questions]: #unresolved-questions

- How to name the config field to not block the future possibilities

# Future possibilities
[future-possibilities]: #future-possibilities

> ### Manifest support

Reviewer 4: ```suggestion
### Support in manifests
```

Add a related field to manifests that the config can override.

### Dependency version unification

> Unlike feature unification, dependency versions are alwayd unified across the

Reviewer 4: ```suggestion
Unlike feature unification, dependency versions are currently always unified across the
```

MSRV-requirements of a low-MSRV package.

We could offer an opt-in to per-package `Cargo.lock` files.  For builds, this
could be implemented similar to `resolver.feature-unification = "package"`.

This could run into problems with
- `cargo update` being workspace-focused
- third-party updating tools

As for the MSRV-case, this would only help if you develop with the latest
versions locally and then have a job that resolves down to your MSRVs.

### Unify features in other settings

[`workspace.resolver = "2"`](https://doc.rust-lang.org/cargo/reference/resolver.html#features) removed unification from the following scenarios
- Cross-platform target unification
- `build-dependencies` / `dependencies` unification
- `dev-dependencies` / `dependencies` unification unless a dev target is enabled

Depending on how we design this, the solution might be good enough to
re-evaluate
[build-target features](https://github.com/rust-lang/rfcs/pull/3374) as we
could offer a way for users to opt-out of build-target unification.

Like with `resolver.incompatible-rust-version`, a solution for this would override the defaults of `workspace.resolver`.

> `cargo hakaro` gives control over `build-dependencies` / `dependencies` unification with

Reviewer 4: ```suggestion
`cargo hakari` gives control over `build-dependencies` / `dependencies` unification with
```

- Feature Name: `feature-unification`
- Start Date: 2024-09-11
- RFC PR: [rust-lang/rfcs#3692](https://github.com/rust-lang/rfcs/pull/3692)

> - Rust Issue: [rust-lang/rust#3692](https://github.com/rust-lang/rust/issues/3692)

Reviewer 5: this is an unrelated issue.

This increases entropy within Cargo and the universe at large.

If someone enables mutually exclusive features in different packages, then `workspace` unification will fail.

> Officially, features are supposedd to be additive., making mutuallyu exclusive features officially unsupported.

Reviewer 3: ```suggestion
Officially, features are supposed to be additive, making mutually exclusive features officially unsupported.
```
